### PR TITLE
yaml_to_mux: Turn yaml_to_mux plugin to optional one [v3]

### PR DESCRIPTION
**Pull request by ldoktor**
Currently the yaml_to_mux plugin is a core plugin without properly
defined requirements. Let's separate it to optional_plugins, define
dependencies and also build it separately in RPM.

v1: https://github.com/avocado-framework/avocado/pull/2016
v2: https://github.com/avocado-framework/avocado/pull/2022

Changes:

```yaml
v2: New commit to log full results on build doc failure
v2: New commit to include API documentation of optional_plugins
v2: Updated documentation
v2: Use python-yaml rather than PyYAML in RPM specfile
v2: Remove main avocado python-yaml dependency
v3: Moved and updated the whole yaml_to_mux documentation section to optional_plugins
v3: Used @clebergnu's suggested heading to optional-plugins-api documentation
v3: added the (optional) `.rst` suffix to match the style (at least at that paragraph)
```

### CODE REVIEW

File: selftests/doc/test_doc_build.py [hunk 1]
```diff
@@ -54,7 +54,9 @@ def test_build_docs(self):
                            'nonexisting document u\'api/test/avocado.utils\'')
         doc_dir = os.path.join(basedir, 'docs')
         process.run('make -C %s clean' % doc_dir)
-        result = process.run('make -C %s html' % doc_dir)
+        result = process.run('make -C %s html' % doc_dir, ignore_status=True)
+        self.assertFalse(result.exit_status, "Doc build reported non-zero "
+                         "status:\n%s" % result)
```
Comment (clebergnu) [inline selftests/doc/test_doc_build.py]: OK.

File: docs/source/index.rst [hunk 2]
```diff
@@ -58,6 +58,7 @@ API Reference
    api/utils/avocado.utils.rst
    api/core/avocado.core.rst
    api/plugins/avocado.plugins.rst
+   api/optional-plugins/index
```
Comment (clebergnu) [inline docs/source/index.rst]: This doesn't match what you described as done: "added the (optional) `.rst` suffix to match the style (at least at that paragraph)".  Actually, it looks like this commit is unchanged from v2.

File: docs/source/index.rst [hunk 3]
```diff
@@ -58,7 +58,7 @@ API Reference
    api/utils/avocado.utils.rst
    api/core/avocado.core.rst
    api/plugins/avocado.plugins.rst
-   api/optional-plugins/index
+   api/optional-plugins/index.rst
```
Comment (clebergnu) [inline docs/source/index.rst]: Oh, here it is.  I guess this was intended to be in 1675c6ba9be81839bfb776deeeb50032ba774850 instead.

File: docs/source/conf.py [hunk 4]
```diff
@@ -101,6 +101,41 @@
...
+        path = os.path.join(optional_plugins_path, path, name)
+        if not os.path.exists(path):
+            continue
+        output_dir = os.path.join(api_optional_plugins_path, name)
+        params = {"api_source_dir": path, "output_dir": output_dir,
+                  "exclude_dirs": ""}
+        process.run(apidoc_template % params)
+        # Remove the unnecessary generated files
+        os.unlink(os.path.join(output_dir, "modules.rst"))
+        optional_plugins_toc.write("\n   %s" % os.path.join(name, name))
```
Comment (clebergnu) [inline docs/source/conf.py]: This brings failures in my local box when building the docs:

```
/home/cleber/src/avocado/avocado/docs/source/api/optional-plugins/avocado_resultsdb/avocado_resultsdb.rst:7: WARNING: autodoc: failed to import module u'avocado_resultsdb'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)
ImportError: No module named avocado_resultsdb
/home/cleber/src/avocado/avocado/docs/source/api/optional-plugins/avocado_robot/avocado_robot.rst:7: WARNING: autodoc: failed to import module u'avocado_robot'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)
ImportError: No module named avocado_robot
/home/cleber/src/avocado/avocado/docs/source/api/optional-plugins/avocado_runner_docker/avocado_runner_docker.rst:7: WARNING: autodoc: failed to import module u'avocado_runner_docker'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)
ImportError: No module named avocado_runner_docker
/home/cleber/src/avocado/avocado/docs/source/api/optional-plugins/avocado_runner_remote/avocado_runner_remote.rst:7: WARNING: autodoc: failed to import module u'avocado_runner_remote'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)
ImportError: No module named avocado_runner_remote
/home/cleber/src/avocado/avocado/docs/source/api/optional-plugins/avocado_runner_vm/avocado_runner_vm.rst:7: WARNING: autodoc: failed to import module u'avocado_runner_vm'; the following exception was raised:
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/sphinx/ext/autodoc.py", line 551, in import_object
    __import__(self.modname)                                                                                                                                                
ImportError: No module named avocado_runner_vm
```

The reason is that the modules are not visible and can not be imported.  A `make develop` solves that, but the problem is that a `make develop` (or similar) is not done on rtd, leading to an also empty content:

http://avocado-framework-ldoktor.readthedocs.io/en/latest/api/optional-plugins/avocado_varianter_yaml_to_mux/avocado_varianter_yaml_to_mux.html

File: docs/source/index.rst (hunk 3 above)
Comment (ldoktor) [inline docs/source/index.rst]: oups, sorry...

File: docs/source/conf.py (hunk 4 above)
Comment (ldoktor) [inline docs/source/conf.py]: well then we need to run in on RTD. As for the local build I think it's expected for users to run `make develop` manually (or we can add basic `make html` target to `Makefile` for convenience, but I doubt many people will have problem with this...

Comment (ldoktor) [inline docs/source/conf.py]: Btw I created an account there but I can't see a way to modify that. Do you know how to do it, or should I simply add that step to `conf.py`?

**Comment by clebergnu** [comment]
[Make yaml2mux optional_plugin](https://trello.com/c/PLBJ9AfN/998-make-yaml2mux-optional-plugin)

**Pull request closed without merging**